[CIR] Add GEP flags to ptr stride op

badumbatish opened the pull request:
From the [comment](https://github.com/llvm/clangir/pull/1844#discussion_r2289392182) on PR review #1844, it seems like we're missing the flags for GEP.

I'm opening the PR to add the flags.

The first commit is just a prototype to gather opinions and reviews to see if I'm heading to the right direction with this.

bcardosolopes commented on clang/include/clang/CIR/Dialect/IR/CIROps.td:
```diff
@@ -386,6 +386,14 @@ def CIR_PtrDiffOp : CIR_Op<"ptr_diff", [Pure, SameTypeOperands]> {
+                     [I32BitEnumAttrCaseNone<"None">,
+                      I32BitEnumAttrCase<"InBounds", 0x1, "inbounds">,
+                      I32BitEnumAttrCase<"NoUnsignedWrap", 0x2, "nuw">,
+                      I32BitEnumAttrCase<"NoSignedWrap", 0x4, "nsw">]> {
```
Are those excluse or they can be set independently? If independent it might make more sense to do something like  `CIR_BinOp`

bcardosolopes commented on clang/lib/CIR/Dialect/IR/CIRDialect.cpp:
```diff
@@ -17,6 +17,7 @@
 #include "clang/CIR/Dialect/IR/CIRTypes.h"
 #include "clang/CIR/Interfaces/CIRLoopOpInterface.h"
 #include "clang/CIR/MissingFeatures.h"
+#include "llvm/ADT/StringExtras.h"
```
maybe you don't need this change

badumbatish replied:
this was to fix this error in the build online
```
2025-08-26T20:30:04.3995412Z FAILED: tools/clang/lib/CIR/Dialect/IR/CMakeFiles/obj.MLIRCIR.dir/CIRDialect.cpp.o 
2025-08-26T20:30:04.4009852Z sccache /opt/llvm/bin/clang++ -DCLANG_EXPORTS -DEXPERIMENTAL_KEY_INSTRUCTIONS -DGTEST_HAS_RTTI=0 -D_DEBUG -D_GLIBCXX_ASSERTIONS -D_GNU_SOURCE -D__STDC_CONSTANT_MACROS -D__STDC_FORMAT_MACROS -D__STDC_LIMIT_MACROS -I/home/gha/actions-runner/_work/clangir/clangir/build/tools/clang/lib/CIR/Dialect/IR -I/home/gha/actions-runner/_work/clangir/clangir/clang/lib/CIR/Dialect/IR -I/home/gha/actions-runner/_work/clangir/clangir/clang/include -I/home/gha/actions-runner/_work/clangir/clangir/build/tools/clang/include -I/home/gha/actions-runner/_work/clangir/clangir/build/include -I/home/gha/actions-runner/_work/clangir/clangir/llvm/include -I/home/gha/actions-runner/_work/clangir/clangir/llvm/../mlir/include -I/home/gha/actions-runner/_work/clangir/clangir/build/tools/mlir/include -gmlt -fPIC -fno-semantic-interposition -fvisibility-inlines-hidden -Werror=date-time -Werror=unguarded-availability-new -Wall -Wextra -Wno-unused-parameter -Wwrite-strings -Wcast-qual -Wmissing-field-initializers -pedantic -Wno-long-long -Wc++98-compat-extra-semi -Wimplicit-fallthrough -Wcovered-switch-default -Wno-noexcept-type -Wnon-virtual-dtor -Wdelete-non-virtual-dtor -Wsuggest-override -Wstring-conversion -Wmisleading-indentation -Wctad-maybe-unsupported -fdiagnostics-color -ffunction-sections -fdata-sections -fno-common -Woverloaded-virtual -Wno-nested-anon-types -O3 -DNDEBUG -std=c++17  -fno-exceptions -funwind-tables -fno-rtti -UNDEBUG -MD -MT tools/clang/lib/CIR/Dialect/IR/CMakeFiles/obj.MLIRCIR.dir/CIRDialect.cpp.o -MF tools/clang/lib/CIR/Dialect/IR/CMakeFiles/obj.MLIRCIR.dir/CIRDialect.cpp.o.d -o tools/clang/lib/CIR/Dialect/IR/CMakeFiles/obj.MLIRCIR.dir/CIRDialect.cpp.o -c /home/gha/actions-runner/_work/clangir/clangir/clang/lib/CIR/Dialect/IR/CIRDialect.cpp
2025-08-26T20:30:04.4036215Z In file included from /home/gha/actions-runner/_work/clangir/clangir/clang/lib/CIR/Dialect/IR/CIRDialect.cpp:49:
2025-08-26T20:30:04.4037848Z /home/gha/actions-runner/_work/clangir/clangir/build/tools/clang/include/clang/CIR/Dialect/IR/CIROpsEnums.cpp.inc:1004:18: error: no member named 'join' in namespace 'llvm'
2025-08-26T20:30:04.4039282Z  1004 |   return ::llvm::join(strs, "|");
2025-08-26T20:30:04.4039690Z       |          ~~~~~~~~^
2025-08-26T20:30:04.4040056Z 1 error generated.
```

bcardosolopes commented on clang/include/clang/CIR/Dialect/IR/CIROps.td:
```diff
@@ -401,15 +409,14 @@ def CIR_PtrStrideOp : CIR_Op<"ptr_stride",[
 
   let assemblyFormat = [{
-    `(` $base `:` qualified(type($base)) `,` $stride `:` qualified(type($stride)) `)`
+    `(` $base `:` qualified(type($base)) `,` $stride `:` qualified(type($stride))(`,` $stride_attr^)?`)`
```
Please add CIR to CIR tests for parsing/printing

badumbatish replied:
yep on it

bcardosolopes reviewed (changes requested):
Nice! This also needs CIRGen changes and lowering to LLVM support, but looks like a good start

tommymcm commented:
It might make sense for `cir.get_element` to get the same treatment. Maybe we should have it inherit from `ptr_stride` to avoid duplicating everything? `get_element` is essentially just a `ptr_stride` with 'better' type information.

badumbatish commented:
is there any historical reason why we didn't consolidate the two?

tommymcm commented:
> is there any historical reason why we didn't consolidate the two?

Nope, I introduced `get_element` recently-ish. It just didn't have any meaningful overlap until this. Other option is having them share a parent (some kind of "ptr_arith" base class). 
This can all be handled in a later PR though.

andykaylor commented:
I like `get_element` as a separate operation, because it does something logically different than `ptr_stride`, but if we can share code between them, we should.

badumbatish commented:
I added lowering and codegen, I added OGCG test initially but was reluctant if i should push it. 

I also didn't remove the missing feature check, i'm not sure how feature ready this is compared to OG.

badumbatish commented:
> Can this be first ported to incubator then mirrored here.

I dont quite follow, this repo is the incubator yes?

badumbatish commented on clang/include/clang/CIR/Dialect/IR/CIROps.td:
```diff
@@ -397,19 +413,19 @@ def CIR_PtrStrideOp : CIR_Op<"ptr_stride",[
 
   let assemblyFormat = [{
-    `(` $base `:` qualified(type($base)) `,` $stride `:` qualified(type($stride)) `)`
+    `(` $base `:` qualified(type($base)) `,` $stride `:` qualified(type($stride))(`,` $noWrapFlags^)?`)`
     `,` qualified(type($result)) attr-dict
```
yeah i can file an issue for that

badumbatish commented on clang/lib/CIR/Lowering/DirectToLLVM/LowerToLLVM.cpp:
```diff
@@ -1023,9 +1023,9 @@ mlir::LogicalResult CIRToLLVMPtrStrideOpLowering::matchAndRewrite(
   rewriter.replaceOpWithNewOp<mlir::LLVM::GEPOp>(
-      ptrStrideOp, resultTy, elementTy, adaptor.getBase(), index);
+      ptrStrideOp, resultTy, elementTy, adaptor.getBase(), index,
+      static_cast<mlir::LLVM::GEPNoWrapFlags>(adaptor.getNoWrapFlags()));
```
sure, i can adjust it

badumbatish commented on clang/include/clang/CIR/Dialect/IR/CIROps.td:
````diff
@@ -397,19 +413,19 @@ def CIR_PtrStrideOp : CIR_Op<"ptr_stride",[
 
     ```mlir
     %3 = cir.const 0 : i32
+    %4 = cir.ptr_stride(%2 : !cir.ptr<i32>, %3 : i32), !cir.ptr<i32>
+
````
I forgot to add the enum in

badumbatish commented on clang/include/clang/CIR/Dialect/IR/CIROps.td:
```diff
@@ -386,6 +386,22 @@ def CIR_PtrDiffOp : CIR_Op<"ptr_diff", [Pure, SameTypeOperands]> {
+  let printBitEnumPrimaryGroups = 1;
+}
+
+def GEPNoWrapFlagsProp : EnumProp<GEPNoWrapFlags> {
+  let defaultValue = interfaceType#"::none";
+}
```
this is just copied verbatim from the LLVM dialect, i'll incorporate your comments to the next push

badumbatish commented on clang/include/clang/CIR/Dialect/IR/CIROps.td:
```diff
@@ -386,6 +386,22 @@ def CIR_PtrDiffOp : CIR_Op<"ptr_diff", [Pure, SameTypeOperands]> {
 //===----------------------------------------------------------------------===//
 // PtrStrideOp
 //===----------------------------------------------------------------------===//
+def GEPNone : I32BitEnumCaseNone<"none">;
```
yep

tommymcm commented on clang/include/clang/CIR/Dialect/IR/CIROps.td:
```diff
@@ -386,6 +386,22 @@ def CIR_PtrDiffOp : CIR_Op<"ptr_diff", [Pure, SameTypeOperands]> {
+def GEPInbounds : BitEnumCaseGroup<"inbounds", [GEPInboundsFlag, GEPNusw]>;
+
+def GEPNoWrapFlags
+    : I32BitEnum<"GEPNoWrapFlags", "::cir::GEPNoWrapFlags",
```
Could we just replace these with `CIR_I32Enum` instead? That way it would match the other `enum`s we use.

badumbatish replied:
i'm more keen on keeping the 1-to-1 mapping between the LLVM dialect and this PR

tommymcm commented on clang/include/clang/CIR/Dialect/IR/CIREnumAttr.td:
```diff
@@ -15,6 +15,11 @@
 
 include "mlir/IR/EnumAttr.td"
 
+class CIR_I32BitEnum<string name, string summary, list<BitEnumCaseBase> cases>
+    : I32BitEnum<name, summary, cases> {
```
Looking at other dialects, I think this should be:
```suggestion
class CIR_I32BitEnumAttr<string name, string summary, list<BitEnumCaseBase> cases>
    : I32BitEnumAttr<name, summary, cases> {
```

tommymcm commented on clang/include/clang/CIR/Dialect/IR/CIROps.td:
```diff
@@ -413,14 +415,18 @@ def CIR_PtrStrideOp : CIR_Op<"ptr_stride",[
 
     %4 = cir.ptr_stride(%2 : !cir.ptr<i32>, %3 : i32), !cir.ptr<i32>
+
+    %4 = cir.ptr_stride(%2 : !cir.ptr<i32>, %3 : i32, inbounds), !cir.ptr<i32>
+
+    %4 = cir.ptr_stride(%2 : !cir.ptr<i32>, %3 : i32, inbounds|nuw), !cir.ptr<i32>
```
```suggestion
    %5 = cir.ptr_stride(%2 : !cir.ptr<i32>, %3 : i32, inbounds), !cir.ptr<i32>

    %6 = cir.ptr_stride(%2 : !cir.ptr<i32>, %3 : i32, inbounds|nuw), !cir.ptr<i32>
```

tommymcm commented on clang/lib/CIR/Lowering/DirectToLLVM/LowerToLLVM.cpp:
```diff
@@ -92,6 +92,25 @@ void walkRegionSkipping(mlir::Region &region,
+/// Convert from a CIR PtrStrideOp kind to an LLVM IR equivalent of GEP.
+mlir::LLVM::GEPNoWrapFlags
+convertPtrStrideKindToGEPFlags(cir::CIR_GEPNoWrapFlags flags) {
+  static std::unordered_map<cir::CIR_GEPNoWrapFlags, mlir::LLVM::GEPNoWrapFlags>
+      mp = {
+          {cir::CIR_GEPNoWrapFlags::none, mlir::LLVM::GEPNoWrapFlags::none},
+          {cir::CIR_GEPNoWrapFlags::inbounds,
+           mlir::LLVM::GEPNoWrapFlags::inbounds},
+          {cir::CIR_GEPNoWrapFlags::inboundsFlag,
+           mlir::LLVM::GEPNoWrapFlags::inboundsFlag},
+          {cir::CIR_GEPNoWrapFlags::nusw, mlir::LLVM::GEPNoWrapFlags::nusw},
+          {cir::CIR_GEPNoWrapFlags::nuw, mlir::LLVM::GEPNoWrapFlags::nuw},
+      };
+  mlir::LLVM::GEPNoWrapFlags x = mlir::LLVM::GEPNoWrapFlags::none;
+  for (auto [key, _] : mp)
+    x = x | mp.at(flags & key);
+  return x;
```
Don't use a `static` map like this, it introduces shared state.

This should just be a series of `if (flags & cir::CIR_GEP...) x |= mlir::LLVM::GEP...;`

xlauko commented:
> > Can this be first ported to incubator then mirrored here.
> 
> I dont quite follow, this repo is the incubator yes?

Yeah sorry, I was jumping in between PRs and missed that this was already in clangir.
